### PR TITLE
Clarify TopK NaN handling behavior in spec

### LINKED ISSUE #4716: How operator TopK should handle nan value?
# Ask a Question
TopK specifications does not specify how to handle nan values. onnxruntime seems to consider them as the top values. Should we ignore them?

### PR DESCRIPTION
**Pull request by ogulcanaydogan**
## Summary

Clarify the NaN handling behavior of the TopK operator in the spec. The reference implementation already treats NaN values as greater than all non-NaN values (consistent with NumPy and IEEE 754 total ordering), but this behavior was not documented in the operator specification.

Changes:
- Add NaN ordering semantics to the TopK operator spec (`utils.cc`)
- Add two test cases: `test_top_k_nan_handling` (largest=1) and `test_top_k_nan_smallest` (largest=0)
- Update reference implementation docstring

This is a spec clarification only, no behavioral changes to the reference implementation.

Fixes #4716

## Test plan

- [x] Verify NaN ordering matches existing reference implementation behavior
- [x] Add `export_top_k_nan_handling` test (NaN with `largest=1`, multiple NaN tiebreaking)
- [x] Add `export_top_k_nan_smallest` test (NaN excluded from smallest-k)
- [ ] CI: `pytest`, `lintrunner`

### CODE REVIEW

**Comment by ogulcanaydogan** [comment]
Hi @justinchuby — since you reviewed the approach in the issue and mentioned it sounds reasonable, would you mind taking a look at this PR when you get a chance? The changes clarify TopK's NaN handling behavior in the spec. Thanks\!

**Comment by ogulcanaydogan** [comment]
Regenerated the auto-generated docs (Operators.md, Changelog.md, TestCoverage.md) — the `Enforce style` check should pass now.

@justinchuby could you add a `topic: operators` or `module: operators` label to resolve the `check-label` failure? Thanks!

**Review by themavik — COMMENTED** [review]
Reviewed the changes — the approach follows the existing patterns and looks correct.

**Review by themavik — COMMENTED** [review]
Reviewed the changes — the approach is clean and follows the existing patterns.

**Comment by ogulcanaydogan** [comment]
Friendly ping - is there anything else needed on this PR? Happy to make changes if needed.

**Comment by ogulcanaydogan** [comment]
Friendly bump — operators-approvers, could you take a look when you have a moment? CI is green and the change is scoped to clarifying TopK NaN handling in the spec. Thanks @justinchuby for the early engagement.

**Review by justinchuby — DISMISSED** [review]

**Comment by justinchuby** [comment]
@onnx/sig-operators for comments

**Comment by fdwr** [comment]
> The reference implementation already treats NaN values as greater than all non-NaN values

Is that *desireable*? 🤔 I know that from a raw bit perspective (treating a float32 as uint32 without the sign), NaN's are greater than infinity and all finite values, but if you had a list of values {42, NaN, 13, NaN} where computation failed with some of the outputs, wouldn't those be the least interesting of the lot? If I was to pick the top two, it seems {42,13} would be more interesting/useful. So, was the behavior of NumPy a side effect of implementation details, or was it intentional for a use case?

> When multiple NaN values are present ...

Interestingly there are multiple flavors of NaN's and [here](https://en.wikipedia.org/wiki/IEEE_754#Total-ordering_predicate) we read "a NaN with a lesser payload is treated as having a lesser absolute value".

**Comment by ogulcanaydogan** [comment]
Good points, both worth thinking about.

On desirability: agreed that NaN at the top of TopK is rarely what users actually want. NaN usually means a computation broke and grabbing it as a top result is mostly noise. The reason for documenting it this way is empirical, not principled: PyTorch's `torch.topk(largest=True)` returns NaNs first, `np.sort` puts them at the end (treats NaN as greatest), and ORT does the same for `largest=1`. Models running across these runtimes silently rely on the convention today, so writing it down at least makes the behavior portable. A `nan_policy` attribute (propagate / ignore / fail) feels cleaner long-term but probably belongs in a separate change.

Worth flagging from local testing: ORT with `largest=0` also returns NaN first, e.g. `[NaN, 1, 2]` for `[1, NaN, 3, 2, NaN, 7]`. That contradicts the "NaN > all" framing for the smallest case, so the spec text needs a caveat there or ORT needs a fix.

On NaN payload ordering: agreed we shouldn't require totalOrder semantics. I'll soften the "lower index first" tiebreaker to say the relative order of NaN values is implementation-defined, so we don't pin implementations to bit-level payload comparisons.

**Comment by ogulcanaydogan** [comment]
Updated the spec text and pushed. Scoped the 'NaN > all' guarantee to largest=1 only, marked NaN-NaN tiebreaker as implementation-defined, removed the largest=0 NaN test since that case is now explicitly implementation-defined.

**Comment by gramalingam** [comment]
> Good points, both worth thinking about.
> 
> On desirability: agreed that NaN at the top of TopK is rarely what users actually want. NaN usually means a computation broke and grabbing it as a top result is mostly noise. The reason for documenting it this way is empirical, not principled: PyTorch's `torch.topk(largest=True)` returns NaNs first, `np.sort` puts them at the end (treats NaN as greatest), and ORT does the same for `largest=1`. Models running across these runtimes silently rely on the convention today, so writing it down at least makes the behavior portable. A `nan_policy` attribute (propagate / ignore / fail) feels cleaner long-term but probably belongs in a separate change.
> 
> Worth flagging from local testing: ORT with `largest=0` also returns NaN first, e.g. `[NaN, 1, 2]` for `[1, NaN, 3, 2, NaN, 7]`. That contradicts the "NaN > all" framing for the smallest case, so the spec text needs a caveat there or ORT needs a fix.
> 
> On NaN payload ordering: agreed we shouldn't require totalOrder semantics. I'll soften the "lower index first" tiebreaker to say the relative order of NaN values is implementation-defined, so we don't pin implementations to bit-level payload comparisons.

This is tricky. I think we have two options: (a) Describe the exact resolution (as in this PR), or (b) Say that it is implementation-defined.

When possible (a) is preferable, since we do want the deployed model inference to have same behavior as the trained model (across different hardware). But if the reality is that different implementations have different behavior (eg., in Nvidia GPU vs some other CPU), then we are better off sticking to (b).

So, my question is: do we know about any implementations that behave differently? I know that it can be challenging to get a comprehensive answer, but if we can check/verify in a few different popular backend implementations, that would help.

**Comment by ogulcanaydogan** [comment]
I ran your example `[1, NaN, 3, 2, NaN, 7]` (k=3) through the three backends I have on hand:

```
PyTorch topk(largest=True):  [NaN, NaN, 7]           # every NaN first
PyTorch topk(largest=False): [1, 2, 3]               # NaNs excluded (treated as large)
ORT largest=1:               [NaN, 7, 3]             # one NaN, then the real maxima
ORT largest=0:               [NaN, 1, 2]             # NaN first here too
numpy sort ascending:        [1, 2, 3, 7, NaN, NaN]  # NaN last
```

So they do diverge, and not only at the edges: for `largest`, PyTorch surfaces every NaN at the top while ORT surfaces one NaN and then the real maxima, for the same input. ORT also isn't internally consistent (`largest=0` returns NaN first too, which contradicts treating NaN as the greatest value). numpy is the only one that consistently sorts NaN as greatest.

That matches your option (b): the NaN result falls out of undefined comparisons, so it depends on the partial-sort algorithm. I think the honest thing is to document the NaN case as implementation-defined rather than prescribe an ordering the spec can't enforce. I'm happy to reword the PR that way: keep the defined ordering for finite values, and state that the presence and relative order of NaNs in the result is implementation-defined. Does that work?
